### ISSUE TITLE
Self built derper failed. Port 443 is OK by default and other ports are not.

### ISSUE THREAD

**Issue by bbhxwl** [tailscale/tailscale#4716]
Self built derper failed. Port 443 is OK by default and other ports are not.

![image](https://user-images.githubusercontent.com/26525599/169509072-f8658ddb-8865-4295-aeac-2b20a4685244.png)
![image](https://user-images.githubusercontent.com/26525599/169509198-a9182e87-9781-41c9-a2eb-9743c4412447.png)

**Comment by DentonGentry** [comment]
I don't know what "failed" means here. It doesn't start? It doesn't answer queries.

**Comment by bbhxwl** [comment]
> I don't know what "failed" means here. It doesn't start? It doesn't answer queries?

![image](https://user-images.githubusercontent.com/26525599/169512031-d827ca8a-c20d-47d0-a1f1-089974dfae78.png)
timeout ，It indicates that the configuration cannot be used, and it seems that he does not use TLS protocol.

**Comment by bbhxwl** [comment]
> I don't know what "failed" means here. It doesn't start? It doesn't answer queries?

![image](https://user-images.githubusercontent.com/26525599/169512414-b7e93b1e-d122-4bb1-b5da-65ee899cef2f.png)

**Comment by bbhxwl** [comment]
> I don't know what "failed" means here. It doesn't start? It doesn't answer queries?

It means that the relay is gone, but it doesn't work

**Comment by DentonGentry** [comment]
From one of the nodes on your tailnet:
```
2022-05-20 18:22:26.3655034 +0800 +0800: derphttp.Client.Recv: connecting to derp-900 (myderp)
2022-05-20 18:22:26.4221851 +0800 +0800: magicsock: [0xc0005581e0] derp.Recv(derp-900): derphttp.Client.Recv connect to region 900 (myderp): tls: first record does not look like a TLS handshake
```

I think there is a Meddler-In-The-Middle (MITM) interfering with your connection to the derper.

**Comment by bbhxwl** [comment]
> From one of the nodes on your tailnet:
> 
> ```
> 2022-05-20 18:22:26.3655034 +0800 +0800: derphttp.Client.Recv: connecting to derp-900 (myderp)
> 2022-05-20 18:22:26.4221851 +0800 +0800: magicsock: [0xc0005581e0] derp.Recv(derp-900): derphttp.Client.Recv connect to region 900 (myderp): tls: first record does not look like a TLS handshake
> ```
> 
> I think there is a Meddler-In-The-Middle (MITM) interfering with your connection to the derper.

Is it because of a bug in the program?
![image](https://user-images.githubusercontent.com/26525599/169524881-2218b7ee-6aed-4389-aa64-756c9336d981.png)

**Comment by bbhxwl** [comment]
![image](https://user-images.githubusercontent.com/26525599/169525551-66ce644b-4edf-46d3-8d11-def7b885dfe7.png)


> From one of the nodes on your tailnet:
> 
> ```
> 2022-05-20 18:22:26.3655034 +0800 +0800: derphttp.Client.Recv: connecting to derp-900 (myderp)
> 2022-05-20 18:22:26.4221851 +0800 +0800: magicsock: [0xc0005581e0] derp.Recv(derp-900): derphttp.Client.Recv connect to region 900 (myderp): tls: first record does not look like a TLS handshake
> ```
> 
> I think there is a Meddler-In-The-Middle (MITM) interfering with your connection to the derper.

**Comment by DentonGentry** [comment]
"tls: first record does not look like a TLS handshake"

There is something between your client and the DERP server which is interfering.

It is not a bug in the program and it is not anything we can help you with. One of the firewalls inside China is intercepting and interfering with your traffic. Other users have reported success in trying different port numbers until they find one which isn't interfered with. I cannot advise what port number you should use.

**Comment by bbhxwl** [comment]
> "tls: first record does not look like a TLS handshake"
> 
> There is something between your client and the DERP server which is interfering.
> 
> It is not a bug in the program and it is not anything we can help you with. One of the firewalls inside China is intercepting and interfering with your traffic. Other users have reported success in trying different port numbers until they find one which isn't interfered with. I cannot advise what port number you should use.

As long as the program I set is not 443, it will not use TLS. Should this be the problem of the program?

**Comment by bbhxwl** [comment]
> From one of the nodes on your tailnet:
> 
> ```
> 2022-05-20 18:22:26.3655034 +0800 +0800: derphttp.Client.Recv: connecting to derp-900 (myderp)
> 2022-05-20 18:22:26.4221851 +0800 +0800: magicsock: [0xc0005581e0] derp.Recv(derp-900): derphttp.Client.Recv connect to region 900 (myderp): tls: first record does not look like a TLS handshake
> ```
> 
> I think there is a Meddler-In-The-Middle (MITM) interfering with your connection to the derper.

![image](https://user-images.githubusercontent.com/26525599/169531401-dfa50324-888f-4480-970a-91ffc50d3712.png)
If it doesn't succeed because it's not a TLS problem, it should be a program problem. I didn't try to connect, and I also use the Japanese server. It's the same. I can't use a non 443 port

**Comment by bbhxwl** [comment]
![image](https://user-images.githubusercontent.com/26525599/169531850-729669c7-3baf-483b-a8e3-58d0e1b2c6ff.png)
I use a Japanese server, but I don't use a Chinese server. If I can't use the Japanese server of tailscale, I can't use it normally. This will not be because of China GFW's problems.

**Comment by bbhxwl** [comment]
![image](https://user-images.githubusercontent.com/26525599/169532451-a57eee0f-e5e1-4a2c-8b43-0b2e6f5ec4d5.png)

**Comment by bbhxwl** [comment]
The above one uses the default 443

**Comment by bbhxwl** [comment]
> "tls: first record does not look like a TLS handshake"
> 
> There is something between your client and the DERP server which is interfering.
> 
> It is not a bug in the program and it is not anything we can help you with. One of the firewalls inside China is intercepting and interfering with your traffic. Other users have reported success in trying different port numbers until they find one which isn't interfered with. I cannot advise what port number you should use.

Custom port. Do you need to apply for a certificate manually?

**Comment by 1itt1eB0y** [comment]
derp服务器有一个[`DERPNode.DERPPort`](https://pkg.go.dev/tailscale.com/tailcfg#DERPNode.DERPPort)的环境变量，你是不是没有配置这个端口导致你的客户端从DERP获取数据的时候，重定向到443，没有使用881端口
官方帖子也有提到这个端口的内容https://tailscale.com/kb/1118/custom-derp-servers/#multiple-nodes-in-a-region
我目前还没自建derp，不能完全确定原因是否出自于此，老哥你先试试

**Comment by 1itt1eB0y** [comment]
兄弟，我试了一下，目前我的这个方案是可行的
DERP运行的时候手动指定所需的域名及其证书和私钥
```
./derper -a :61255 -certdir /certdir -certmode manual -stun-port 61255 -hostname exp.com -http-port -1
```
证书会自动在对应文件夹下寻找 `exp.com.crt`，私钥则会寻找`exp.com.key`，把Let's Encrypt的证书和私钥复制重命名一份就可以
服务器到此就正常启动了，然后去修改你的ACL使其对应你在上面命令里设置的域名
注意：derper目前对泛域名证书好像不支持，泛域名证书无法正常识别，所以要申请一个单域名证书

我猜测主要原因还是在于Let's Encrypt自动申请证书下发的时候需要验证443端口来证明你对域名的所有权，但是迁移到其他端口之后无法正常验证443端口所有权，导致对证书的认证失效，无法正常握手

---

Bro, I tried it, and here is my solution
Manually specify the required domain name and its certificate and private key when DERP is running
````
./derper -a :61255 -certdir /certdir -certmode manual -stun-port 61255 -hostname exp.com -http-port -1
````
The certificate will automatically look for `exp.com.crt` in the corresponding folder, and the private key will look for `exp.com.key`, just copy and rename the certificate and private key of Let's Encrypt
The server starts normally at this point, and then modify your ACL to correspond to your domain name as you set above the command
Note: Derper does not seem to support pan-domain certificates at present, and pan-domain certificates cannot be recognized normally, so you need to apply for a single-domain certificate

I assume, the main reason is that when Let's Encrypt automatically applies for certificate issuance, it needs to verify port 443 to prove your ownership of the domain name. However, after migrating to other ports, the ownership of port 443 cannot be verified normally, resulting in invalid certification of the certificate and no normal handshake.

**Comment by DentonGentry** [comment]
> Custom port. Do you need to apply for a certificate manually?

A port other than 443 will need to use a manually supplied certificate. LetsEncrypt only allows port 443.

Make sure to use a relatively recent build, a problem with manual certificates was fixed in early August. https://github.com/tailscale/tailscale/pull/5336

**Comment by DentonGentry** [comment]
I think this is resolved, once manual certificates are added.

**Comment by chenxudong2020** [comment]
> > “TLS：第一条记录看起来不像 TLS 握手”
> > 您的客户端和 DERP 服务器之间存在干扰。
> > 这不是程序中的错误，也不是我们可以为您提供的任何东西。中国境内的防火墙之一正在拦截和干扰您的流量。其他用户报告说，他们成功地尝试了不同的端口号，直到找到一个不受干扰的端口号。我不能建议你应该使用什么端口号。
> 
> 只要我设置的程序不是 443，它就不会使用 TLS。这应该是程序的问题吗？

Using custom certificates for non-standard ports requires configuration

**Comment by izeroo** [comment]
> > Custom port. Do you need to apply for a certificate manually?
> 
> A port other than 443 will need to use a manually supplied certificate. LetsEncrypt only allows port 443.
> 
> Make sure to use a relatively recent build, a problem with manual certificates was fixed in early August. #5336

From what I know, port is not related with ssl certificate since we don't have to specify port when issueing a cert.
But it is indeed somehow related to port. I used a non-standard port for https with a wildcard cert issued by Let's Encrypt. When I use Edge to test that URL it just works fine. But if I use curl to test that URL curl will fail on checking common name. And my android client can't connect to the relay while windows just works fine. So I think it depends on implementation.
I spent quite some time on this wierd problem since derp(d)? gave zero warning about this.

In short, if you are using a wildcard cert with custom port, do re-issue a new non-wildcard cert.
